T6013: Add support for AuthorizedPrincipalsFile to trusted_user_ca_key

takehaya opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
This PR complements [vyos/vyos-1x#4234](https://github.com/vyos/vyos-1x/pull/4234) by adding support for `AuthorizedPrincipalsFile`. With this enhancement, SSH login using CA certificates and arbitrary user principals is now fully supported.

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
* https://vyos.dev/T6013

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->
* https://github.com/vyos/vyos-1x/pull/4234
* https://github.com/vyos/vyos-documentation/pull/1641

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->

## Proposed changes
<!--- Describe your changes in detail -->

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->

## Smoketest result
<!-- Provide the output of the smoketest
```
$ /usr/libexec/vyos/tests/smoke/cli/test_xxx_feature.py
test_01_simple_options (__main__.TestFeature.test_01_simple_options) ... ok
```
-->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

takehaya committed:
ssh: T6013: support SSH AuthorizedPrincipalsFile in use with trusted-user-ca-key

Thisc omplements commit e7cab89f9f81 ("T6013: Add support for configuring
TrustedUserCAKeys in SSH service with local and remote CA keys"). It introduces
a new CLI node per user to support defining the authorized principals used by
any given PKI certificate. It is now possible to associate SSH login users with
their respective principals.

Authored-by: Takeru Hayasaka <hayatake396@gmail.com>

dmbaturin reviewed (approved):
I do not have any objections against this feature, and I cannot see any logic issues, either. It's a straightforward and useful extension of https://github.com/vyos/vyos-1x/pull/4234

c-po commented on interface-definitions/service_ssh.xml.in:
```diff
@@ -281,6 +281,25 @@
               #include <include/pki/ca-certificate.xml.i>
+              <tagNode name="bind-user">
+                <properties>
+                  <help>user-name</help>
```
In order to make this feature properly work this most likely corresponds to a username defined already under: `set system login user` - so this node misses a completionHelper for lazy people like me:

```suggestion
                  <help>Username for principal list</help>
                  <completionHelp>
                    <path>system login user</path>
                  </completionHelp>                  
```

dmbaturin replied:
I agree, @c-po's idea is better than adding it to `service ssh`.

c-po commented on interface-definitions/service_ssh.xml.in:
```diff
@@ -281,6 +281,25 @@
+                <children>
+                  <leafNode name="principal">
+                    <properties>
+                      <help>principal-name</help>
```
```suggestion
                      <help>Principal listed in user certificate</help>
```

c-po reviewed (changes requested):
Thank you for the PR and sorry for the late feedback.

This is a good addition to the TrustedUserCAKeys feature. But I do not like the CLI design of this. As this implementation mainly focuses on the fact to allow a principal for a given user - and that file is rendered to the users home directory - this is no more different then how we manage SSH keys - except the file path is different.

I would request a change in CLI from:
`vyos@vyos# set service ssh trusted-user-ca-key bind-user vyos principal AwesomeCertFooBar`
to
`vyos@vyos# set system login user cpo authentication principal AwesomeCertFooBar`

Thus we have everything under the same place in our configuration and make the users life a little less miserable.

Please also extend the PRs section for `How to test` and `Smoketest results`

c-po committed:
ssh: T6013: move principal name to "system login user <name> authentication"

We already support using per-user SSH public keys for system authentication.
Instead of introducing a new CLI path to configure per-user principal names,
we should continue using the existing CLI location and store the principal
names alongside the corresponding SSH public keys.

set system login user <name> principal <principal>

The certificate used for SSH authentication contains an embedded principal
name, which is defined under this CLI node. Only users with matching principal
names are permitted to log in.

c-po committed:
pki: T6013: add proper dependencies for SSH CA

We need to establish proper dependencies on "system login" and "pki ca" for
the SSH subsystem. If the CA is updated or user principal names are modified,
we must also ensure that the SSH daemon is restarted accordingly.

c-po committed:
ssh: T6013: rename trusted-user-ca-key -> truster-user-ca

The current implementation for SSH CA based authentication uses "set service
ssh trusted-user-ca-key ca-certificate <foo>" to define an X.509 certificate
from "set pki ca <foo> ..." - fun fact, native OpenSSH does not support X.509
certificates and only runs with OpenSSH ssh-keygen generated RSA or EC keys.

This commit changes the bahavior to support antive certificates generated using
ssh-keygen and loaded to our PKI tree. As the previous implementation
did not work at all, no migrations cript is used.

c-po reviewed (approved):
```
cpo@LR1.wue3:~$ /usr/libexec/vyos/tests/smoke/cli/test_service_ssh.py 
test_ssh_default (__main__.TestServiceSSH.test_ssh_default) ... ok
test_ssh_dynamic_protection (__main__.TestServiceSSH.test_ssh_dynamic_protection) ... ok
test_ssh_login (__main__.TestServiceSSH.test_ssh_login) ... ok
test_ssh_multiple_listen_addresses (__main__.TestServiceSSH.test_ssh_multiple_listen_addresses) ... ok
test_ssh_ndcpp (__main__.TestServiceSSH.test_ssh_ndcpp) ... ok
test_ssh_pubkey_accepted_algorithm (__main__.TestServiceSSH.test_ssh_pubkey_accepted_algorithm) ... ok
test_ssh_single_listen_address (__main__.TestServiceSSH.test_ssh_single_listen_address) ... ok
test_ssh_trusted_user_ca (__main__.TestServiceSSH.test_ssh_trusted_user_ca) ... ok
test_ssh_vrf_multi (__main__.TestServiceSSH.test_ssh_vrf_multi) ... ok
test_ssh_vrf_single (__main__.TestServiceSSH.test_ssh_vrf_single) ... ok

----------------------------------------------------------------------
Ran 10 tests in 47.562s

OK
```
